utils/host_utils.cpp

```diff
@@ -39,11 +47,677 @@ int V_ex, Vh_ex;
+    mg_eig_use_dagger[i] = QUDA_BOOLEAN_FALSE;
+    mg_eig_use_poly_acc[i] = QUDA_BOOLEAN_TRUE;
+    mg_eig_poly_deg[i] = 100;
+    mg_eig_amin[i] = 1.0;
+    mg_eig_amax[i] = -1.0; // use power iterations
+
+    setup_ca_basis[i] = QUDA_POWER_BASIS;
+    setup_ca_basis_size[i] = 4;
+    setup_ca_lambda_min[i] = 0.0;
+    setup_ca_lambda_max[i] = -1.0; // use power iterations
+
+    coarse_solver_ca_basis[i] = QUDA_POWER_BASIS;
+    coarse_solver_ca_basis_size[i] = 4;
+    coarse_solver_ca_lambda_min[i] = 0.0;
+    coarse_solver_ca_lambda_max[i] = -1.0;
+
+    strcpy(mg_vec_infile[i], "");
+    strcpy(mg_vec_outfile[i], "");
+  }
+}
+
+void setQudaDefaultMgSolveTypes()
+{
+  for (int i = 0; i < QUDA_MAX_MG_LEVEL; i++) {
+    if (coarse_solve_type[i] == QUDA_INVALID_SOLVE) coarse_solve_type[i] = solve_type;
+    if (smoother_solve_type[i] == QUDA_INVALID_SOLVE) smoother_solve_type[i] = QUDA_DIRECT_PC_SOLVE;
+  }
+}
+
+void performanceStats(double *time, double *gflops)
+{
+  auto mean_time = 0.0;
+  auto mean_time2 = 0.0;
+  auto mean_gflops = 0.0;
+  auto mean_gflops2 = 0.0;
+  // skip first solve due to allocations, potential UVM swapping overhead
+  for (int i = 1; i < Nsrc; i++) {
+    mean_time += time[i];
+    mean_time2 += time[i] * time[i];
+    mean_gflops += gflops[i];
+    mean_gflops2 += gflops[i] * gflops[i];
+  }
+
+  auto NsrcM1 = Nsrc - 1;
+
+  mean_time /= NsrcM1;
+  mean_time2 /= NsrcM1;
+  auto stddev_time = NsrcM1 > 1 ? sqrt((NsrcM1 / ((double)NsrcM1 - 1.0)) * (mean_time2 - mean_time * mean_time)) :
+                                  std::numeric_limits<double>::infinity();
+  mean_gflops /= NsrcM1;
+  mean_gflops2 /= NsrcM1;
+  auto stddev_gflops = NsrcM1 > 1 ? sqrt((NsrcM1 / ((double)NsrcM1 - 1.0)) * (mean_gflops2 - mean_gflops * mean_gflops)) :
+                                    std::numeric_limits<double>::infinity();
+  printfQuda("%d solves, with mean solve time %g (stddev = %g), mean GFLOPS %g (stddev = %g) [excluding first solve]\n",
+             Nsrc, mean_time, stddev_time, mean_gflops, stddev_gflops);
+}
+
+// Spinor construction routines
+void constructRandomSpinorSource(void *v, int nSpin, int nColor, QudaPrecision precision, const int *const x,
+                                 quda::RNG &rng)
+{
+  quda::ColorSpinorParam param;
+  param.v = v;
+  param.nColor = nColor;
+  param.nSpin = nSpin;
+  param.setPrecision(precision);
+  param.create = QUDA_REFERENCE_FIELD_CREATE;
+  param.fieldOrder = QUDA_SPACE_SPIN_COLOR_FIELD_ORDER;
+  param.nDim = 4;
+  param.siteSubset = QUDA_FULL_SITE_SUBSET;
+  param.siteOrder = QUDA_EVEN_ODD_SITE_ORDER;
+  param.location = QUDA_CPU_FIELD_LOCATION; // DMH FIXME so one can construct device noise
+  for (int d = 0; d < 4; d++) param.x[d] = x[d];
+  quda::cpuColorSpinorField spinor_in(param);
+  quda::spinorNoise(spinor_in, rng, QUDA_NOISE_UNIFORM);
+}
+
+void constructStaggeredTestSpinorParam(quda::ColorSpinorParam *cs_param, const QudaInvertParam *inv_param,
```

Comment:
should we move the staggered specific function in their own file?

Reply:
b8e9ba5ee